source/common/common/packed_struct.h

```diff
@@ -0,0 +1,137 @@
+#pragma once
+
+#include <algorithm>
+#include <limits>
+#include <type_traits>
+
+#include "source/common/common/assert.h"
+#include "source/common/common/utility.h"
+#include "source/common/protobuf/utility.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+
+/**
+ * Helper class to pack elements of the same type into a struct like
```

Comment:
Can you write a little summary of the implementation details? It's a bit magical and it would be good to have a summary to avoid having to look through all the code to figure out how it works.

Reply:
PTAL, and let me know if the summary makes sense.